File: ruby/prime-factors/prime_factors.rb
module PrimeFactors
  def self.of(n, factors = [])
    least_factor = (2..n).find { |i| n % i == 0 }

    if least_factor
      factors << least_factor
      return of(n / least_factor, factors)
    else
      return factors
    end
  end
end

File: javascript/prime-factors/prime-factors.js
export const primeFactors = (n, factors = []) => {
    let r = range(2, n);

    let i, least_factor;

    while (i = r.next().value) {
        if (n % i == 0) {
            least_factor = i;
            break;
        }
    }

    if (least_factor) {
        factors.push(least_factor);
        return primeFactors(parseInt(n / least_factor), factors);
    } else {
        return factors;
    }
};

function* range(begin, end) {
    let i = begin;

    while (i <= end)
        yield i++;
}
